clickhouse/base/sslsocket.cpp

```diff
@@ -143,42 +178,42 @@ SSLSocket::SSLSocket(const NetworkAddress& addr, const SSLParams & ssl_params, S
     if (!ssl)
         throw std::runtime_error("Failed to create SSL instance");
 
+    std::unique_ptr<ASN1_OCTET_STRING, decltype(&ASN1_OCTET_STRING_free)> ip_addr(a2i_IPADDRESS(addr.Host().c_str()), &ASN1_OCTET_STRING_free);
+
     HANDLE_SSL_ERROR(ssl, SSL_set_fd(ssl, handle_));
     if (ssl_params.use_SNI)
         HANDLE_SSL_ERROR(ssl, SSL_set_tlsext_host_name(ssl, addr.Host().c_str()));
 
+    HANDLE_SSL_ERROR(ssl, SSL_set1_host(ssl, addr.Host().c_str()));
+    if (ssl_params.host_flags != -1)
+        SSL_set_hostflags(ssl, ssl_params.host_flags);
+
+    if (ssl_params.configuration.size() > 0)
+        configureSSL(ssl_params.configuration, ssl);
+
     SSL_set_connect_state(ssl);
     HANDLE_SSL_ERROR(ssl, SSL_connect(ssl));
     HANDLE_SSL_ERROR(ssl, SSL_set_mode(ssl, SSL_MODE_AUTO_RETRY));
-    auto peer_certificate = SSL_get_peer_certificate(ssl);
-
-    if (!peer_certificate)
-        throw std::runtime_error("Failed to verify SSL connection: server provided no certificate.");
 
-    if (const auto verify_result = SSL_get_verify_result(ssl); verify_result != X509_V_OK) {
+    if (const auto verify_result = SSL_get_verify_result(ssl); !ssl_params.skip_verification && verify_result != X509_V_OK) {
         auto error_message = X509_verify_cert_error_string(verify_result);
         throw std::runtime_error("Failed to verify SSL connection, X509_v error: "
                 + std::to_string(verify_result)
                 + " " + error_message
-                + "\nServer certificate: " + getCertificateInfo(peer_certificate));
+                + "\nServer certificate: " + getCertificateInfo(SSL_get_peer_certificate(ssl)));
     }
 
-    if (ssl_params.use_SNI) {
-        auto hostname = addr.Host();
-        char * out_name = nullptr;
-
-        std::unique_ptr<ASN1_OCTET_STRING, decltype(&ASN1_OCTET_STRING_free)> addr(a2i_IPADDRESS(hostname.c_str()), &ASN1_OCTET_STRING_free);
-        if (addr) {
-            // if hostname is actually an IP address
-            HANDLE_SSL_ERROR(ssl, X509_check_ip(
-                    peer_certificate,
-                    ASN1_STRING_get0_data(addr.get()),
-                    ASN1_STRING_length(addr.get()),
-                    0));
-        } else {
-            HANDLE_SSL_ERROR(ssl, X509_check_host(peer_certificate, hostname.c_str(), hostname.length(), 0, &out_name));
-        }
-    }
+    // Host name verification is done by OpenSSL itself, however if we are connecting to an ip-address,
+    // no verification is made, so we have to do it manually.
+    // Just in case if this is ever required, leave it here commented out.
```

Comment:
is SSL_set_verify SSL_VERIFY_PEER not needed? 
https://www.openssl.org/docs/man1.1.1/man3/SSL_set_verify.html

at also appears in the examples in https://github.com/iSECPartners/ssl-conservatory/blob/1449f3ae918623699947e24f388c531b30ad2c87/openssl/test_client.c#L115-L119

https://wiki.openssl.org/index.php/Hostname_validation

Reply:
Yep, makes sense to have some safe defaults